File: leetcode/no0216_Combination_Sum_III.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
const DBG = false;
var combinationSum3 = function (k, n) {
    if (DBG) console.log("INPUT:", k, n);
    let nums = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    function dfs(nums, targetSum, currntSum, currentCombination, startedIndex, answer) {
        if (currntSum == targetSum && currentCombination.length == k) {
            if (DBG) console.log("!!PUSH:", currentCombination);
            answer.push(currentCombination.slice(0));
            return;
        }
        for (let i = startedIndex; i < nums.length; i++) {
            let num = nums[i];
            currentCombination.push(num);
            dfs(nums, targetSum, currntSum + num, currentCombination, i + 1, answer);
            currentCombination.pop();
        }
    }
    let ans = [];
    let curr = [];
    dfs(nums, n, 0, curr, 0, ans);
    if (DBG) console.log("OUTPUT:", ans);
    return ans;
};

//combinationSum3(3, 7);// Output: [[1,2,4]]
combinationSum3(3, 9);// Output: [[1,2,6],[1,3,5],[2,3,4]]

File: leetcode/no0216_Combination_Sum_III.go
package main

import "fmt"

func combinationSum3(k int, n int) [][]int {

	candidates := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}

	var backtrack func(
		candidates []int,
		comb []int,
		currSum int,
		targetSum int,
		startedIndex int,
		ans [][]int,
	) [][]int
	backtrack = func(
		candidates []int,
		comb []int,
		currSum int,
		targetSum int,
		startedIndex int,
		ans [][]int,
	) [][]int {
		if currSum == targetSum && len(comb) == k {
			clone := make([]int, len(comb))
			copy(clone, comb)

			ans = append(ans, clone)
			return ans
		}

		if currSum > targetSum {
			return ans
		}

		for i := startedIndex; i < len(candidates); i++ {

			num := candidates[i]
			if num > targetSum {
				continue
			}
			comb = append(comb, num)
			ans = backtrack(
				candidates,
				comb,
				currSum+num,
				targetSum,
				i+1,
				ans,
			)
			comb = comb[:len(comb)-1]
		}

		return ans
	}

	answer := [][]int{}
	curr := []int{}

	answer = backtrack(
		candidates,
		curr,
		0,
		n,
		0,
		answer,
	)

	return answer
}

func main() {

	k := 3
	n := 7
	result := combinationSum3(k, n)
	fmt.Println("OUTPUT:", result)

}
